092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+```
+
+Where the user wants to enable self-healing for some specific anomalies, they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.
+
+### Strimzi Notifier
+
+Cruise Control provides `AnomalyNotifier` interface which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface. 
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc.
+The `SelfHealingNotifier` class can then be further used to implement your own notifier
+The other custom notifier offered by Cruise control are also based upon the `SelfHealingNotifier`.
+
+We will be using our own notifier implementation called `StrimziNotifer`(set via the `anomaly.notifier.class` config).
+Strimzi will configure Cruise Control to always use the `StrimziNotifier` to notify regarding the anomalies and take action on them.
+
+For effective alerts and monitoring, we will be making use of Kubernetes events to signal to users what operations Cruise Control is performing automatically.
+The `StrimziNotifier` will override the `alert()` method of the `SelfHealingNotifier` and will trigger and publish events whenever an anomaly is detected.
+These events would publish information regarding the anomaly like the anomalyID, the type of anomaly which was detected etc.
+The events would help the user to understand the anomalies that are present, when the anomalies appeared in their cluster and also would help them know what is going on in their clusters.
+
+The event published would look like this:
+```yaml
+- action: <action>
+  apiVersion: v1
+  eventTime: "2024-12-19T08:06:21.314140Z"
+  firstTimestamp: null
+  involvedObject:
+    kind: Pod
+    name: <cruise-control-pod-name>
+    namespace: myproject
+  kind: Event
+  lastTimestamp: null
+  message: Cruise Control is fixing the detected anomaly
+  metadata:
+    creationTimestamp: "2024-12-19T08:06:21Z"
+    generateName: cruise-control-event
+    name: my-cluster-34c41626-0add-47a3-9b79-d6ee29d5d81f  # my-cluster-<Anomaly ID>
+    namespace: myproject
+    resourceVersion: "101070"
+    uid: 80892c7d-e3a1-44ad-84dc-2d88e15a7ff0
+  reason: Anomaly was detected in the cluster GOAL_VIOLATION # Detected Anomaly type
+  reportingComponent: cruise-control
+  reportingInstance: cruise-control
+  source: {}
+  type: Normal
+```
+
+The `StrimziNotifier` class is going to reside in a separate module in the operator repository called the `strimzi-cruise-control-notifier` so that a seperate jar for the notifier can be built for inclusion in the Cruise Control container image.
```

Comment:
* Why should it be in the operators repo? Seems like it has nothing to do with the operator? So Why not a separate repository?
* There is no Cruise Control container image, Cruise Control is in the Kafka images.